src/gui/qgsabstractrelationeditorwidget.cpp

```diff
@@ -241,7 +241,7 @@ QgsFeatureIds QgsAbstractRelationEditorWidget::addFeature( const QgsGeometry &ge
     // n:m Relation: first let the user create a new feature on the other table
     // and autocreate a new linking feature.
     QgsFeature finalFeature;
-    if ( !vlTools->addFeature( mNmRelation.referencedLayer(), QgsAttributeMap(), geometry, &finalFeature ) )
+    if ( !vlTools->addFeature( mNmRelation.referencedLayer(), QgsAttributeMap(), geometry, &finalFeature, this, false, true ) )
```

Comment:
By passing `showModal=false` the add feature dialog is opened not blocking with `open` instead of `exec`.
This causes `addFeature` to return immediately and all consecutively logic around `finalFeature` fails. The first feature id returned by `QgsAbstractRelationEditorWidget::addFeature` will not be valid.
In multiedit mode, only the first feature will get the values edited in the add feature dialog.

I don't know how to solve this without changing the return type of `QgsAbstractRelationEditorWidget::addFeature` or make the dialog modal (blocking with `exec`) again.

@signedav do you maybe have an idea?

Reply:
As far as I see, you don't need to change the `QgsAbstractRelationEditorWidget::addFeature` and just don't rely on the `finalFeature`. Instead waiting for the return by a signal. I know it's not that nice, but I would really like not making it modal because this brings lot's of other trouble.

Reply:
Yes I also thing the dialog should stay non-modal. How can I wait on a signal without blocking the gui thread? Some sort of wait & process events, do you know of an example where we do that?

Reply:
But then it means `QgsAbstractRelationEditorWidget::addFeature` will not be able to return the correct Ids, because `vlTools->addFeature` returns immediately while the dialog is still open.

Reply:
Ah now I get it. Hm... With `QEventLoop` the GUI thread gets blocked, so we cannot proceed on the other dialog? 

But anyway I feel medium comfortable with a wait. What are the alternatives? Only to change the return type as you mentioned. This breaks the API - but did it once work on an LTR at all`? Can you check that?
And yes it needs changes on multi edit (what I belief you know better than me).